### ISSUE TITLE
@rjsf/material-ui breaks tree shaking

### ISSUE THREAD

**Issue by mdarocha** [rjsf-team/react-jsonschema-form#2748]
### Prerequisites

- [x] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/);
- [x] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [ ] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net/n1k0/f2y3fq7L/6/) or a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### Description

Importing `@rjsf/material-ui` into a project bundled with [esbuild]() causes bundle size to increase a lot.

This seems to be because of https://github.com/rjsf-team/react-jsonschema-form/blob/master/packages/material-ui/src/Theme5/Mui5Context.tsx#L11, which causes the entire `@mui/icons-material` package to be included in the bundle.

Bundle size with `@rjsf/material-ui` included: 6,9M
Bundle size without `@rjsf/material-ui`: 2,4M

### Steps to Reproduce

1. Include `@rjsf/material-ui` in a project
2. Bundle it using a tree-shaking enabled bundler (like esbuild)
3. Compare bundle sizes before and after `@rjsf/material-ui` inclusion

#### Expected behavior

The icons should be tree-shaken and only icons used by the application should be included in the bundle.

#### Actual behavior

All icons from the `@mui/icons-material` package are included in the bundle.

### Version

```
        "@rjsf/core": "^4.0.1",
        "@rjsf/material-ui": "^4.0.1",
```

**Comment by epicfaace** [comment]
> This seems to be because of https://github.com/rjsf-team/react-jsonschema-form/blob/master/packages/material-ui/src/Theme5/Mui5Context.tsx#L11, which causes the entire `@mui/icons-material` package to be included in the bundle.

Do you know how we can import that package and still preserve tree-shaking?

**Comment by epicfaace** [comment]
cc @heath-freenome

**Comment by jdbence** [comment]
It's the same issue when using MUI5 with unneeded material-ui. [https://github.com/rjsf-team/react-jsonschema-form/blob/d2957cd9fc70128f462f918b9f32c47a857b43ac/packages/material-ui/src/Theme/MaterialUIContext.tsx#L9](https://github.com/rjsf-team/react-jsonschema-form/blob/d2957cd9fc70128f462f918b9f32c47a857b43ac/packages/material-ui/src/Theme/MaterialUIContext.tsx#L9)

**Comment by heath-freenome** [comment]
@epicfaace I don't think it is really possible... Maybe it does make sense to keep the two in separate packages

**Comment by mdarocha** [comment]
Yes, what @heath-freenome said - it seems that when a module is imported using `require()`, tree shaking does not work and the entire imported package is bundled.

[MUI recommends using path imports](https://mui.com/guides/minimizing-bundle-size/#option-1) when importing in libraries to avoid such a situation.

For experimentation, I tested changing the `require` to use path imports (in `dist/material-ui.esm.js`):
```js
var mui$1;
var icons$1;

try {
  mui$1 = /*#__PURE__*/require('@mui/material');
  icons$1 = {
    Add: /*#__PURE__*/require('@mui/icons-material/Add'),
    ArrowUpward: /*#__PURE__*/require('@mui/icons-material/ArrowUpward'),
    ArrowDownward: /*#__PURE__*/require('@mui/icons-material/ArrowDownward'),
    Error: /*#__PURE__*/require('@mui/icons-material/Error'),
    Remove: /*#__PURE__*/require('@mui/icons-material/Remove')
  };
} catch (err) {
  if (process.env.NODE_ENV !== 'PRODUCTION') {
    console.log(err);
  }
}
```

Results:
- Unchanged package: 5,9M. Bundle analysis show the entire @mui/icons-material package included
- Changed: 1,8M. Bundle analysis shows only used icons from the @mui/icons-material package

The `@mui/material` import can be changed in the same way, to allow proper tree shaking and import of only used components.

**Comment by rollercodester** [comment]
I have a PR ready for this. Switched to using dynamic imports instead of require:

https://github.com/rjsf-team/react-jsonschema-form/pull/2751

**Comment by rollercodester** [comment]
> @mdarocha do you know why using `require` breaks tree-shaking?

The current code requires in entire packages (`require('@material-ui/core')`, `require('@mui/material')`, `require('@mui/icons-material')` therefore, transpiler engines can't shake out anything. Tree-shaking is determined based on the import statements and not inferred from usage in the code.

**Comment by epicfaace** [comment]
Thanks. Why can't we use option 2 (
https://mui.com/guides/minimizing-bundle-size/#option-2) again?

--
Ashwin Ramaswami

**Comment by rollercodester** [comment]
That's what we would like to do, but the issue is we need conditional imports (dynamic) because the material dependencies are peer and the consumers may not have them both installed (thus you have to check at time of import). An example is my current project is 100% MUI5, so we will not have MUI4 installed.

**Comment by epicfaace** [comment]
Thanks for the explanation!

**Comment by heath-freenome** [comment]
> I have a PR ready for this. Switched to using dynamic imports instead of require: #2751
> 
> I've published a version to my personal NPM in case anyone wants to test it out in their project: https://www.npmjs.com/package/@rollercodester/material-ui

@rollercodester Given some of the changes in the test snapshots, I'm worried that the changes that were made aren't working properly since v5 classes are showing in in v4 snapshots. It also looks like the solution around using /*#__PURE__*/ proposed above may be a better choice?

**Comment by rollercodester** [comment]
@heath-freenome I have no experience with the `/*#__PURE__*/` directive, but if it does the trick, I'm all for it. Less code is better code (most of the time).

**Comment by rollercodester** [comment]
A clarification I should make on this PR. This does not enable tree-shaking, rather it introduces code-splitting. So, the bundle for the MUI dependencies are split out into their own bundle file which is only downloaded when the consumer code calls it. The reason this makes a drastic difference is because I am *explicitly* using named imports to import each component individually. What this means is instead of all of the mui/icon library being downloaded...only the components used with `react-jsonschema-form` are bundled/downloaded. It's not tree-shaking, but since we must have dynamic logic because v5 and v5 are side-by-side, then code-splitting is the next best thing.

I'm very curious as to how the `/*__PURE__%/` directive can be used and if we can indeed get tree-shaking working with that. If so, then that is the awesome way to go!

**Comment by heath-freenome** [comment]
@rollercodester @mdarocha I may try to meld the two approaches when I have time this weekend... Get a better approach to type handling that is in @rollercodester along with trying the `/*#__PURE__*/` directive to get tree-shaking

**Comment by rollercodester** [comment]
I spent a little more time on this after remembering that the MUI project is structured so that each component can be required on its own. This negates the need for code-splitting, rather each component just needs to be required individually.

I took a stab at another PR (#2754), making sure that all of the components are individually required and also properly typed (i.e. no more use of *any*). Additionally, I abstracted context usage into a hook so that the component consumption is much easier to maintain in the future.

I've verified that this results in a 68% (6.34mb) reduction in the resulting bundle. The previous (yucky) PR has been closed in lieu of #2754.

@[heath-freenome](https://github.com/heath-freenome), feel free to ignore if you still want to dig in this weekend. I did this mainly to ease my own sanity as it was *bugging* me.
